share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -537,38 +539,195 @@ private static ShareGroupOffset merge(ShareGroupOffset soFar, ShareUpdateValue n
+        Queue<PersisterStateBatch> batchQueue = new LinkedList<>(
+            mergeBatches(
+                pruneBatches(
+                    getSortedList(batchesSoFar),
+                    startOffset
+                )
+            ));
+
+        // will take care of overlapping batches
+        List<PersisterStateBatch> modifiedNewBatches = new ArrayList<>(
+            mergeBatches(
+                pruneBatches(
+                    getSortedList(newBatches),
+                    startOffset
+                )
+            ));
+
+        for (PersisterStateBatch batch : modifiedNewBatches) {
+            for (int i = 0; i < batchQueue.size(); i++) {
+                PersisterStateBatch cur = batchQueue.poll();
+                // cur batch under inspection has no overlap with the new one
+                // we will need to add to our result
+                if (batch.lastOffset() < cur.firstOffset() || batch.firstOffset() > cur.lastOffset()) {
+                    batchQueue.add(cur);
+                    continue;
+                }
+
+                // Covers cases where we need to create a new interval
+                // from the current one such that they do not
+                // overlap with the new one.
+                // Following cases will not produce any new records so need not be handled.
+                // cur:     ____      ______     ______       _____
+                // new: ________      ______     _________  _________
+                
+                
+                // covers
+                // cur:   ______   _____   _____   ______
+                // batch:   ___    _____   ___        ___
+                if (batch.firstOffset() >= cur.firstOffset() && batch.lastOffset() <= cur.lastOffset()) {
+                    // extra batch needs to be created
+                    if (batch.firstOffset() > cur.firstOffset()) {
+                        batchQueue.add(
+                            new PersisterStateBatch(
+                                cur.firstOffset(),
+                                batch.firstOffset() - 1,
+                                cur.deliveryState(),
+                                cur.deliveryCount()
+                            )
+                        );
+                    }
+
+                    // extra batch needs to be created
+                    if (batch.lastOffset() < cur.lastOffset()) {
+                        batchQueue.add(
+                            new PersisterStateBatch(
+                                batch.lastOffset() + 1,
+                                cur.lastOffset(),
+                                cur.deliveryState(),
+                                cur.deliveryCount()
+                            )
+                        );
+                    }
+                } else if (batch.firstOffset() < cur.firstOffset() && batch.lastOffset() < cur.lastOffset()) {
+                    // covers
+                    //  ______    
```

Comment:
I like the little diagrams. I suggest adding the `cur:` and `batch:` to these on lines 630 and 640 too.